letsencrypt/models/letsencrypt.py

```diff
@@ -78,91 +75,144 @@ def validate_domain(self, domain):
+            """Goes through the altnames and searches for names containing *.
+            If found then all the altnames with the same components are
+            collapsed.
+            """
+            for name in altnames:
+                if '*' in name:
+                    _label_count = name.split('.')
+                    for _name in [
+                            x for x in altnames
+                            if name in x and
+                            name.split('.') == _label_count]:
+                        altnames.remove(_name)
+        altnames = _sort_domains_by_components(altnames)
+        altnames = _clean_same_label_domains(altnames)
+        if altnames:
+            valid_domains = [x.value for x in altnames] + [domain]
         else:
-            _logger.info('no command defined for reloading webserver, please '
-                         'do it manually in order to apply new certificate')
+            valid_domains = [domain]
+        _logger.info(
+            'Making CSR for the following domains: ' + str(valid_domains))
+        return crypto_util.make_csr(
+            open(domain_key_file).read(), valid_domains)
+
+    def _respond_challenge_http(self, challenge, account_key):
+        """Respond to the HTTP challenge by writing the file to serve.
+        """
+        token = base64.urlsafe_b64encode(challenge.token)
+        challenge_file = join(_get_challenge_dir(), '%s' % token)
+        with open(challenge_file, 'wb') as challenge_file:
+            # remove the trailing equals sign (=)
+            challenge_file.write(token + '.' + base64.urlsafe_b64encode(
+                account_key.thumbprint()).rstrip('='))
+
+    def _respond_challenge_dns(self, challenge, domain):
+        """Respond to the DNS challenge by creating the DNS record on the provider
+        """
+        ir_config_parameter = self.env['ir.config_parameter']
+        try:
+            letsencrypt_dns_function = 'letsencrypt_' + \
+                ir_config_parameter.get_param('letsencrypt_dns_provider') + \
+                '_support'
+            getattr(self, letsencrypt_dns_function)(challenge, domain)
+        except (AttributeError, TypeError):
+            _logger.error(
+                'No letsencrypt_dns_function found. '
+                'Install the appropriate support module for your DNS provider '
+                'and create a letsencrypt_$DNS_PROVIDER_support function in '
+                'this model.')
+
+    @api.model
+    def _call_cmdline(
+            self, cmdline, loglevel=logging.INFO, raise_on_result=True):
+        process = subprocess.Popen(
+            cmdline, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        stdout, stderr = process.communicate()
+        if process.returncode:
+            raise exceptions.Warning(_(
+                'Error calling %s: %d') % (cmdline[0], process.returncode),
+                ' '.join(cmdline))
+
+    @api.model
+    def letsencrypt_shell_support(self, *args):
```

Comment:
let's prepend those functions with an underscore too just for good measure.
And you should fix the interface, from my understanding, you pass the challenge and the domain, so list those here too. You'll also need to expose those to the shell command, otherwise there's no chance for it to do something useful. I'd propose to extend `_call_cmdline` to accept an optional parameter `env` which you pass to `Popen`. Then here you can pass the arguments as environment variables to the script.